File: app/src/main/java/by/anegin/telegram_contests/core/data/source/impl/AssetsDataSource.java
package by.anegin.telegram_contests.core.data.source.impl;

import android.content.Context;
import android.graphics.Color;
import by.anegin.telegram_contests.core.data.model.Chart;
import by.anegin.telegram_contests.core.data.model.Column;
import by.anegin.telegram_contests.core.data.model.Data;
import by.anegin.telegram_contests.core.data.source.DataSource;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.*;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

public class AssetsDataSource implements DataSource {

    private final Context context;
    private final String assetName;

    public AssetsDataSource(Context context, String assetName) {
        this.context = context;
        this.assetName = assetName;
    }

    @Override
    public Data getData() throws IOException {
        String jsonString;
        try (InputStream is = context.getAssets().open(assetName)) {
            Reader reader = new InputStreamReader(is, Charset.forName("UTF-8"));
            Writer buffer = new StringWriter();
            char[] buf = new char[8192];
            int chars = reader.read(buf);
            while (chars >= 0) {
                buffer.write(buf, 0, chars);
                chars = reader.read(buf);
            }
            jsonString = buffer.toString();
        }
        try {
            return parseJsonString(jsonString);
        } catch (JSONException e) {
            throw new IOException(e);
        }
    }

    private Data parseJsonString(String jsonString) throws JSONException {
        List<Chart> charts = new ArrayList<>();

        JSONArray dataJson = new JSONArray(jsonString);
        for (int i = 0; i < dataJson.length(); i++) {
            JSONObject chartJson = dataJson.optJSONObject(i);
            if (chartJson == null) continue;

            JSONArray columnsJson = chartJson.optJSONArray("columns");
            JSONObject typesJson = chartJson.optJSONObject("types");
            JSONObject namesJson = chartJson.optJSONObject("names");
            JSONObject colorsJson = chartJson.optJSONObject("colors");
            if (columnsJson == null || typesJson == null || namesJson == null || colorsJson == null) continue;
            if (columnsJson.length() == 0 || typesJson.length() == 0 || namesJson.length() == 0 || colorsJson.length() == 0)
                continue;

            Column.X x = null;
            List<Column.Line> lines = new ArrayList<>();
            for (int j = 0; j < columnsJson.length(); j++) {

                JSONArray columnJson = columnsJson.optJSONArray(j);
                if (columnJson == null) continue;

                String columnId = columnJson.optString(0);
                String columnType = typesJson.optString(columnId);
                if (columnType == null) continue;

                int valuesCount = columnJson.length() - 1;
                if ("x".equals(columnType)) {

                    long[] columnValues = new long[valuesCount];
                    for (int v = 0; v < valuesCount; v++) {
                        columnValues[v] = columnJson.optLong(v + 1);
                    }

                    x = new Column.X(columnId, columnValues);

                } else if ("line".equals(columnType)) {

                    String columnName = namesJson.optString(columnId);
                    if (columnName == null) columnName = "No name";

                    int columnColor;
                    try {
                        columnColor = Color.parseColor(colorsJson.optString(columnId));
                    } catch (IllegalArgumentException e) {
                        columnColor = Color.BLACK;
                    }

                    long[] columnValues = new long[valuesCount];
                    for (int v = 0; v < valuesCount; v++) {
                        columnValues[v] = columnJson.optLong(v + 1);
                    }

                    lines.add(new Column.Line(columnId, columnName, columnColor, columnValues));
                }
            }

            if (x != null && lines.size() > 0) {
                charts.add(new Chart(x, lines));
            }
        }

        return new Data(charts);
    }

}

File: app/src/main/java/by/anegin/telegram_contests/data/source/impl/AssetsDataSource.kt
package by.anegin.telegram_contests.data.source.impl

import android.content.Context
import android.graphics.Color
import by.anegin.telegram_contests.data.source.DataSource
import by.anegin.telegram_contests.data.model.Chart
import by.anegin.telegram_contests.data.model.Column
import by.anegin.telegram_contests.data.model.Data
import org.json.JSONArray
import org.json.JSONException
import java.io.IOException

class AssetsDataSource(
    private val context: Context,
    private val assetName: String
) : DataSource {

    @Throws(IOException::class)
    override fun getData(): Data {
        return context.assets.open(assetName).use {
            val jsonString = it.bufferedReader().readText()
            try {
                parseJsonString(jsonString)
            } catch (e: JSONException) {
                throw IOException(e)
            }
        }
    }

    @Throws(JSONException::class)
    private fun parseJsonString(jsonString: String): Data {
        val charts = ArrayList<Chart>()

        val dataJson = JSONArray(jsonString)
        for (i in 0 until dataJson.length()) {
            val chartJson = dataJson.optJSONObject(i) ?: continue

            val columnsJson = chartJson.optJSONArray("columns") ?: continue
            val typesJson = chartJson.optJSONObject("types") ?: continue
            val namesJson = chartJson.optJSONObject("names") ?: continue
            val colorsJson = chartJson.optJSONObject("colors") ?: continue

            if (columnsJson.length() == 0 || typesJson.length() == 0
                || namesJson.length() == 0 || colorsJson.length() == 0
            ) continue

            var x: Column.X? = null
            val lines = ArrayList<Column.Line>()
            for (j in 0 until columnsJson.length()) {
                val columnJson = columnsJson.optJSONArray(j) ?: continue

                val columnId = columnJson.optString(0)
                val columnType = typesJson.opt(columnId) ?: continue

                val valuesCount = columnJson.length() - 1

                if (columnType == "x") {

                    val columnValues = LongArray(valuesCount)
                    for (v in 0 until valuesCount) {
                        columnValues[v] = columnJson.optLong(v + 1)
                    }

                    x = Column.X(columnId, columnValues)

                } else if (columnType == "line") {

                    val columnName = namesJson.optString(columnId) ?: "No name"

                    val columnColor: Int = try {
                        Color.parseColor(colorsJson.optString(columnId))
                    } catch (e: IllegalArgumentException) {
                        Color.BLACK
                    }

                    val columnValues = LongArray(valuesCount)
                    for (v in 0 until valuesCount) {
                        columnValues[v] = columnJson.optLong(v + 1)
                    }

                    lines.add(Column.Line(columnId, columnName, columnColor, columnValues))
                }
            }

            if (x != null && lines.isNotEmpty()) {
                charts.add(Chart(x, lines))
            }
        }

        return Data(charts)
    }

}